FIX Allow default value and other options in DBFields

GuySartorelli opened the pull request:
> [!IMPORTANT]
> There are intentionally two commits. DO NOT SQUASH

1. first commit:
    - Don't obstruct options array
    - Make sure default values are passed to requireField
    - Fix options where necessary
1. second commit:
    - Deprecate `DBField->arrayValue` and the corresponding getter and setter, which have never been used from what I can tell since their introduction in framework 2.3
    - Deprecate `DBEnum->default` and the corresponding getter and setter, which are redundant to `getDefaultValue()` and `setDefaultValue()`.

Doing this in a minor because _theoretically_ someone could be using the `setArrayValue()` functionality to do something that somehow breaks stuff when being passed in through options. That's extremely unlikely, especially since `arrayValue` is never used anywhere in core, nor has it even been used since its introduction in framework 2.3 from what I can tell, not even in the postgresql, mssql, or sqlite modules... but better safe than sorry.

Note that the docs don't specify anything about default values for date,time,year etc fields - it explicitly refers to ints (and I'll update that to include float/double/etc), and strings. For this reason (and because the utility would be minimal) I haven't done anything about pushing default values through for date, datetime, year, etc.

> [!NOTE]
> It's probably a good idea to read the changelog details before merging this as it changes some things that might not be obvious just from looking at the code changes.


## Issue

- https://github.com/silverstripe/silverstripe-framework/issues/11873

GuySartorelli committed:
FIX Allow default value and other options in DBFields

- Don't obstruct options array
- Make sure default values are passed to requireField
- Fix options where necessary

GuySartorelli committed:
API Deprecate redundant methods in DBFields

- `DBEnum::getDefault()` and the setter are redundant to
  `getDefaultValue()` and its setter.
- `DBField::getArrayValue()` and its setter have never been used since
  their introduction in framework 2.3 from what I can tell. They do
literally nothing, except get in the way in DBSchemaManager.

GuySartorelli commented on src/ORM/Connect/DBSchemaManager.php:
```diff
@@ -412,7 +412,6 @@ public function requireTable(
                 if ($pos !== false) {
                     //If so, remove it and store that info separately
                     $arrayValue = substr($fieldSpec, $pos);
-                    $fieldSpec = substr($fieldSpec, 0, $pos);
```
Removing this line keeps the options array in the field spec which is passed through into the injector. Without this the options _never_ get to the constructor.

GuySartorelli commented on src/ORM/FieldType/DBEnum.php:
```diff
@@ -77,21 +79,21 @@ public function __construct(
             if ($default && !is_int($default)) {
                 if (in_array($default, $enum)) {
-                    $this->setDefault($default);
+                    $options['default'] = $default;
```
Just include the default in the options array so `DBString` can manage it

GuySartorelli commented on src/ORM/FieldType/DBField.php:
```diff
@@ -65,6 +66,8 @@ abstract class DBField extends ModelData implements DBIndexable, FieldValidation
+     *
+     * @deprecated 6.2.0 Will be removed without equivalent functionality to replace it in a future major release.
      */
     protected $arrayValue;
```
This hasn't been used at all from what I can tell since its introduction in framework 2.3 - not even used in any of the additional db modules like postgresql or mssql and _certainly_ not used for anything in modern kitchen sink.

The only thing the commit message that adds it said is "adds support for array value" (or something very similar, I'm paraphrasing) which... doesn't really mean anything for most DBField types? What's an array value for a string or for an int?

It's just confusing and pointless - and the code to pass it in (in `DBSchemaManager`) is why the options weren't being used correctly.

GuySartorelli commented on src/ORM/FieldType/DBString.php:
```diff
@@ -30,8 +30,13 @@ abstract class DBString extends DBField
     {
-        $this->options['nullifyEmpty'] = true;
-        $this->options['default'] = '';
+        $options = is_array($options) ? $options : [];
+        if (!array_key_exists('nullifyEmpty', $options)) {
+            $options['nullifyEmpty'] = true;
+        }
+        if (!array_key_exists('default', $options)) {
+            $options['default'] = '';
+        }
```
Only set these default options if explicit option values for these keys aren't being passed in

GuySartorelli commented on src/ORM/FieldType/DBString.php:
```diff
@@ -49,10 +54,13 @@ public function setOptions(array $options = []): static
     {
         parent::setOptions($options);
 
-        if (array_key_exists('nullifyEmpty', $options ?? [])) {
+        if (array_key_exists('nullifyEmpty', $options)) {
             $this->options['nullifyEmpty'] = (bool) $options['nullifyEmpty'];
         }
-        if (array_key_exists('default', $options ?? [])) {
+        if (array_key_exists('default', $options)) {
+            if ($this->getNullifyEmpty() && $this->isEmptyValue($options['default'])) {
+                $options['default'] = null;
+            }
```
If we're nullifying empty, the empty string default should be null. Otherwise we're _not_ nullifying empty.

Another solution would be to pass the default through `prepValueForDB()` (either in `setDefaultValue()` or in the various `getFieldSpec()` implementations) - happy to go that route if you prefer.

GuySartorelli commented on src/ORM/FieldType/DBString.php:
```diff
@@ -104,6 +112,11 @@ public function prepValueForDB(mixed $value): array|string|null
         return '';
     }
 
+    protected function isEmptyValue(mixed $value): bool
```
Same logic used in `prepValueForDB()` and `isEmptyValue()`.

GuySartorelli commented on src/ORM/FieldType/DBText.php:
```diff
@@ -59,8 +59,8 @@ public function getFieldSpec(): string|array
             'collate' => $collation,
-            'default' => $this->defaultVal,
-            'arrayValue' => $this->arrayValue
+            'default' => $this->getDefaultValue(),
+            'arrayValue' => $this->arrayValue,
```
`defaultVal` isn't even a thing

GuySartorelli replied:
Note that MySQL doesn't actually support default values for TEXT columns - but if trying to pass one through it'll give a clear exception on dev/build which I think is better than just silently not setting the default.

GuySartorelli commented on tests/php/ORM/DBFieldTest.php:
```diff
@@ -424,12 +425,14 @@ public function testSaveIntoRespectsSetters()
             DBDecimal::class => 0.0,
             DBInt::class => 0,
             DBFloat::class => 0.0,
+            DBClassName::class => $dataObjectClasses[array_key_first($dataObjectClasses)],
```
Note that the default value passed into the database was _always_ this value (see https://github.com/silverstripe/silverstripe-framework/issues/11165). This test was just checking the result of `getDefaultValue()` but the value being passed to the database came from `DBClassName::getDefault()`.

So now `getDefaultValue()` will give you _the actual default value_ which is correct, so updating this test is just now reflecting the accurate value.

GuySartorelli commented on tests/php/ORM/DBVarcharTest.php:
```diff
@@ -15,16 +16,34 @@ class DBVarcharTest extends SapphireTest
         $obj = new DBVarcharTest\TestObject();
+        /** @var DBVarchar $dbField */
+        $dbField = $obj->dbObject('Title');
+        $this->assertTrue($dbField->getNullifyEmpty());
```
We weren't testing `nullifyEmpty` before.

emteknetnz reviewed (changes requested):
Chatted offline - default for string columns will change from null to blank string, which seems like a breaking change e.g. projects having conditional logic on `$value === ""`. Since this is changing the MySQL schema, this also applies to raw SQL queries as well as the ORM

GuySartorelli commented on src/ORM/FieldType/DBString.php:
```diff
@@ -49,10 +54,10 @@ public function setOptions(array $options = []): static
-        if (array_key_exists('default', $options ?? [])) {
+        if (array_key_exists('default', $options)) {
             $this->setDefaultValue($options['default']);
         }
```
Note that previously `$this->options['default']` was being set to empty string - but `setDefaultValue()` wasn't being set! This put it in a weird situation where the default value in the database was an empty string (somehow - still not sure how that got there before), but calling `getDefaultValue()` would return `null`.

GuySartorelli commented:
Changed it but then changed it back - after comparing with an environement that doesn't have this PR.

```php
$x = new MyDataObject();
$id = $x->write();
$xFromDb = MyDataObject::get()->byID($id);

// the following gives null for e.g. regular `Varchar` column in both situations
$x->Title;
$x->dbObject('Title')->value;
$x->dbObject('Title')->defaultValue;
$xFromDb->Title;

// the following gives an empty string in both scenarios
$x->dbObject('Title')->getOptions()['default'];

// the following gives an empty string before this PR, but null after it:
$x->dbObject('Title')->setDefaultValue('')->defaultValue;
```

Calling `setDefaultValue()` could be done in injector config, e.g:

```yaml
SilverStripe\Core\Injector\Injector:
  MySpecialVarchar:
    class: SilverStripe\ORM\FieldType\DBVarchar
    properties:
      DefaultValue: ""
```

Why anyone would ever do that though, I have no idea. It's a workaround for setting a default value, but you'd use it to set an actual value not an empty string (and I doubt anyone's even doing that).

GuySartorelli commented:
I'll update the changelog to remove the false note about changing the default from empty string, since in the db schema it was always null.

emteknetnz commented on src/ORM/FieldType/DBClassName.php:
```diff
@@ -20,8 +22,13 @@ public function getFieldSpec(): string|array
+     */
     public function getDefault(): string
     {
+        Deprecation::notice('6.2.0', 'Use getDefaultValue() instead.');
```
Why are we getting rid of this method?

GuySartorelli replied:
Because we should be using `getDefaultValue()`. We don't need two methods to get the default value - and we especially don't want them to provide different results.

emteknetnz reviewed (changes requested):
Note that for 

```
// the following gives an empty string before this PR, but null after it:
$x->dbObject('Title')->setDefaultValue('')->defaultValue;
```

In my testing I actually got empty string after this PR, which is good as it means this tiny "regression" does not seem like it happened

There is one thing that I'm not sure if it's an issue or not

```php
// app/src/MyDataObject.php

use SilverStripe\ORM\DataObject;

class MyDataObject extends DataObject
{
    private static $table_name = 'MyDataObject';

    private static $db = [
        'Title' => 'Varchar(15, ["default" => "XYZ", "nullifyEmpty" => false])'
    ];
}
```

Where I run this, I bet null, despite `nullifyEmpty` being `false` - possibly unrelated (or even expected?)

```php
$x = new MyDataObject();
$id = $x->write();

$xFromDb = MyDataObject::get()->byID($id);
$xFromDb->Title = '';

$xFromDb->write();
$xFromDb = MyDataObject::get()->byID($id);

var_dump($xFromDb->Title);die; // null
```

GuySartorelli commented:
> Where I run this, I bet null, despite nullifyEmpty being false - possibly unrelated (or even expected?)

I have fixed this and added a test for it.

This was a result of `writeToManipulation()` passing through the result of `nullValue()` if `exists()` was false.
I have retained that - i.e. `nullValue()` still returns `null` no matter what (null is null), and `exists()` still returns false for empty strings (empty strings are still empty values) - but I've overridden `writeToManipulation()` so that it correctly passes an empty string to the database is `nullifyEmpty` is false.

This won't break any existing code because you couldn't set `nullifyEmpty` to false before (unless you were doing some weird injector shenanigans - but anyone doing that would have wanted it to nullify empty so this is an improvement for them too).
The thing I was worried about when we spoke about this in person just now was if I changed the return value of `nullValue()` or `exists()` which I haven't done.
